src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -133,152 +154,11 @@ internal static partial class DependentTypeFinder
-                c => FindTransitivelyImplementingStructuresClassesAndInterfacesWorkerAsync(type, solution, projects, c),
-                cancellationToken);
-        }
-
-        private static Task<ImmutableArray<INamedTypeSymbol>> FindTransitivelyImplementingStructuresClassesAndInterfacesWorkerAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            IImmutableSet<Project> projects,
-            CancellationToken cancellationToken)
-        {
-            return FindDerivedAndImplementingTypesAsync(
-                type, solution, projects, transitive: true, cancellationToken);
-        }
-
-        /// <summary>
-        /// Used for implementing the Inherited-By relation for progression.
-        /// </summary>
-        public static Task<ImmutableArray<INamedTypeSymbol>> FindImmediatelyDerivedAndImplementingTypesAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            CancellationToken cancellationToken)
-        {
-            return FindTypesFromCacheOrComputeAsync(
-                type, solution, projects: null,
-                cache: s_typeToImmediatelyDerivedAndImplementingTypesMap,
-                findAsync: c => FindDerivedAndImplementingTypesAsync(
-                    type, solution, projects: null, transitive: false, c),
-                cancellationToken: cancellationToken);
-        }
-
-        private static Task<ImmutableArray<INamedTypeSymbol>> FindDerivedAndImplementingTypesAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            IImmutableSet<Project> projects,
-            bool transitive,
-            CancellationToken cancellationToken)
-        {
-            // Only an interface can be implemented.
-            if (type?.TypeKind == TypeKind.Interface)
-            {
-                bool metadataTypeMatches(SymbolSet s, INamedTypeSymbol t)
-                    => TypeDerivesFrom(s, t, transitive) || TypeImplementsFrom(s, t, transitive);
-
-                return FindTypesAsync(type, solution, projects,
-                    metadataTypeMatches: metadataTypeMatches,
-                    sourceTypeImmediatelyMatches: ImmediatelyDerivesOrImplementsFrom,
-                    shouldContinueSearching: s_isInterfaceOrNonSealedClass,
-                    transitive: transitive,
-                    cancellationToken: cancellationToken);
-            }
-
-            return SpecializedTasks.EmptyImmutableArray<INamedTypeSymbol>();
-        }
-
         private static async Task<ImmutableArray<INamedTypeSymbol>> FindTypesAsync(
             INamedTypeSymbol type,
             Solution solution,
             IImmutableSet<Project> projects,
-            Func<SymbolSet, INamedTypeSymbol, bool> metadataTypeMatches,
-            Func<SymbolSet, INamedTypeSymbol, bool> sourceTypeImmediatelyMatches,
+            TypeMatches typeMatches,
```

Comment:
instead of needing to pass two functions which were effectively identical (except that metadataTypeMatches would transitively match, and sourceTypeIMmediatelyMatches would not), we now pass in the same function, but have the caller pass in if it should be transitive or not.  The caller knows if it's working with metadata or source, so it can trivially pass the right thing in.